### PR TITLE
Allow users to have multiple openHAB instances in one openHAB-cloud account

### PR DESCRIPTION
**Pull request by FlorianSW**
Solves #29

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -782,26 +805,47 @@ function addAndroidRegistration(req, res) {
...
 app.all('/cometVisu*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
+app.all('/:uuid/proxy*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
 app.all('/proxy*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
+app.all('/:uuid/greent*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
 app.all('/greent*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
+app.all('/:uuid/jquery.*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
 app.all('/jquery.*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
+app.all('/:uuid/classicui/*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
 app.all('/classicui/*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
+app.all('/:uuid/paperui/*', ensureRestAuthenticated, preassembleBody, setOpenhab, proxyRouteOpenhab);
```
Comment (FlorianSW) [inline app.js]: This works great, as long as the responded body contains relative URLs, as they'll use the same prefixed UUID, however, if the body has absolute URL (like all /rest/ requests), they'll end being executed against the first openHAB returned by findOne :/ I haven't found a good way to work around this in openHAB-cloud only. My first idea was to simply rewrite the whole returned content and replace the proxied things to prepent the UUID to it. However, this doesn't work as easy as it sounds, as this normally changes the content length, which will be send before the content is retrieved, so we would need to rewrite the proxy-handling a bit to achieve this.

And this is the point where I would like to hear better ideas :D I don't think, that this is the best solution we can have, so I hope someone else has a better idea :)

File: views/index.ejs [hunk 2]
```diff
@@ -95,30 +95,25 @@
...
+                <% if (openhabs.length > 0) { %>
+                    <% for (var i = 0; i < openhabs.length; i++) { %>
+                        <div class="span5">
+                            <h3><%=openhabs[i].name || 'Unknown openHAB'%><span class="openhabstatus <%=openhabs[i].status%>"></span></h3>
+                            <div class="divider"><span></span></div>
+                            <% if (openhabs[i].status !== 'online') { %>
+                                This openHAB instance is not online. Please make sure that it is running or recheck your <a href="/openhab">openHAB settings</a>.
+                            <% } else { %>
+                                <% if (openhabs[i].openhabVersion && openhabs[i].openhabVersion.split('.')[0] === '2') { %>
+                                You are using openHAB 2.x. <a href="<%=baseurl%>/<%=openhabs[i].uuid%>/start/index">Click here to access your openHAB's dashboard</a>
```
Comment (FlorianSW) [inline views/index.ejs]: the UUID should be set only, if multi instances are enabled.

**Comment by FlorianSW** [comment]
Ok, this is now ready. As far as I can see, I covered all functions and features (at least the ones I can see) and they're working with both modes (multi-instance enabled and not enabled). I'm very interested in any review and comment! :)

**Comment by marziman** [comment]
Cool stuff! Ok this is important and needs testing accordingly! Thanks. If I am done I with the rest I will take a look..

**Comment by FlorianSW** [comment]
Yep, this change should be tested excessive to make sure it's working properly. I would even propose to have a testing server first before this feature is activated (there's a feature flag for it) in production, just to give people time to test it with their environment.

**Comment by marziman** [comment]
@FlorianSW,
I havent forget this..please give me some time, since we have the openhab general assembly on Saturday and I got to speak there.

**Comment by marziman** [comment]
@FlorianSW ,
also soon gonna check this. Can we agree to avoid big PRs? Cause this makes the Review more difficult and if possible we should keep them small (of course only if possible).
This one needs my full testing power and attention, so I would ask for little time :-)
Thanks for your good help!

**Comment by FlorianSW** [comment]
> also soon gonna check this. Can we agree to avoid big PRs?

Yes, of course! I usually try this, however, some features need a lot of changes and during the implementation, a lot of code is changed. It will then take a huge amount of time to split the change. However, I totally agree with you, big PRs isn't the solution. I'll try to do that in the future. (Unfortunately, there's no way to have dependencies between PRs, which makes this a bit more difficult :/).

**Comment by digitaldan** [comment]
This sounds great, before i test it I was curious about a few things. 
 
1. How would a end device (say our android client) target a specific OH instance?
2. How do services like IFTTT work in this case (with multiple OH now)?
3. How are the UI's changed?  For example https://myopenhab.org/items ? 
4. How are the users (https://myopenhab.org/users) associated to a OH instance?
5. Any other "gotchas" I should be aware of?

**Comment by FlorianSW** [comment]
Hi, there're some things, I probably have to work on, too, so I'll answer your questions and indicate, where more work is needed :)

# How would a end device (say our android client) target a specific OH instance?
-> As long as the device is using the sitemaps API, it will get a merged array of sitemaps for all OH instances with the OH ID in the URL, e.g.: myopenhab.com/1234567/rest/sitemaps/whatever. With this URL, if the client requests something, the ID will be set in a cookie and used for future requests, until the client requests another sitemap from another OH instance. So, openhab-cloud sends the correct URLs for the sitemaps rest api for targeting the correct OH instance. I tested that with Android, and it's working fine, however, I've no apple device to test :/
# How do services like IFTTT work in this case (with multiple OH now)?
I haven't checked IFTTT, sorry, missed that completely. However, I've not setup IFTTT, too, so probably you can check this for me? :]
# How are the UI's changed? For example
That's a ToDo, I haven't checked this :(
# How are the users (https://myopenhab.org/users) associated to a OH instance?
The OH instance is associated to the account, so a user should've access to all OH instances of one account. However, I haven't tested this, too :(
# Any other "gotchas" I should be aware of?
More or less one: The feature is behind a feature flag, so you would need to enable it first to test it :)

I know, that here are a lot of things, that may or may not work already and can be improved. That's why I've put the feature behind a feature flag. I'm not sure, if the things, which need to be done now, too, should be done in this PR (it's already a quite big one), or if we can do that in future PRs until we remove the feature flag? :)

**Comment by FlorianSW** [comment]
@digitaldan Is there an easy way to get a test service published for IFTTT to use it with a local openhab-cloud instance? I probably just don't get it how I can test the openHAB service locally with this PR applied :( How do you this? Thanks for your help!

**Comment by digitaldan** [comment]
>s there an easy way to get a test service published for IFTTT to use it with a local openhab-cloud instance?

I think I can setup a staging instance of our account with IFTTT, this would allow us to have a second IFTTT applet to test with, I would also need to add you to the admins list there so your account could access the applet to test

**Comment by digitaldan** [comment]
is https://ifttt.com/openhab_staging visible to you?

**Comment by digitaldan** [comment]
So this sounds like a WIP for now, correct?  Before we merge, even if its behind a feature flag,  it would be nice to know how the above questions will be answered.  Once we have a plan there, then i think we can look at starting to commit the code.

**Comment by FlorianSW** [comment]
> is https://ifttt.com/openhab_staging visible to you?

Unfortunately now :(

> So this sounds like a WIP for now, correct?
More or less. IFTTT is a bit new for me, however, I think I can get it running and figure out how this commit can work with IFTTT. I probably will need some help here, but I'm sure we can find a solution. Once this is resolved, I think we can take a look to the other points (which doesn't sound like a blocker to me, as there's no definition what would happens, so we can be free in defining it, I think). What do you think (tl;dr: IFTTT needs to be fixed within this PR, otherwise it can't be merged :P)?

**Comment by digitaldan** [comment]
>I think we can take a look to the other points 

There are some issues here I would like to get more comfortable with before we move forward.  One issue is that I don't think cookies are going to work here.  Take the rest API,  we have many integrations, mobile clients, web services, etc... that will not use a cookie. So how do we support this.  My guess is this is not a show stopper, but I would still like to know what changes then are we starting to impose (like prefixing the URL, having a default endpoint, GUI's to manage this, etc..).  

Could you layout a roadmap for how these would all look?  I am trying to avoid lots of code going into the project that might not get turned on if we run into significant issues, scope or  time constraints.   

The other concern I have is that accessing my.openhab as a proxy starts to look far different then accessing the OH server directly.  The fact that both are API identical (for the OH rest api) in terms of the URI, integrations mostly just work  between the two.

Does that make sense?

**Comment by marziman** [comment]
@FlorianSW:
Are you still on this? Are you going to finish this, before tackling other stuff? I think this is a good one!

**Comment by kaikreuzer** [comment]
Sorry to come late to the party - I just have one concern:

>  As long as the device is using the sitemaps API, it will get a merged array of sitemaps for all OH instances with the OH ID in the URL, e.g.: myopenhab.com/1234567/rest/sitemaps/whatever. 

This sounds pretty nasty/dangerous. I would expect the different instances to simply have different urls, e.g. https://home.myopenhab.org/123 and https://home.myopenhab.org/456 - and those would be cleanly separated from each other, just like http://192.168.0.2 and http://192.168.0.3 are. Maybe it would even work to work with subdomains, e.g. https://id1.home.myopenhab.org and https://id2.home.myopenhab.org?

**Comment by FlorianSW** [comment]
@kaikreuzer Using a special URL part, such as Google does it for multiple open tabs in GMail for multiple accounts:
mail/u/0/ -> the first logged in user
mail/u/1/ -> the second logged in user
and so on, was my first idea, too. However, the biggest problem is, that openHAB-cloud would require the response from the openHAB instance to be rewritten, as the openHAB instance does not know, that the frontend-URL is the openHAB URL and that this URL starts with /u/whatever or whatever indicator we use. This is more or less a fuzzy logic, as this would need to be string replace over all content (including changing the content length) in all files, including JavaScript files, e.g..

I tried that already and from what experience I had, this is fairly difficult to do in openhab-cloud alone. There might be a better working solution, if we implement something in openHAB itself, that it can recognize, that it is accessed from a reverse proxy and need to adapt the URLs in the response correctly.

**Comment by FlorianSW** [comment]
I've started from scratch (by taking some of the changes from this PR over, however), and rebuild this feature a bit. The old work is still available in the branch [issue/29_old](https://github.com/FlorianSW/openhab-cloud/commits/issue/29_old). This now does not rely on a cookie anymore, which needs to be set, however, it does not support all features anymore, like I mentioned in the commit message of the last commit:

* Rest API does not work as expected, it will show the swagger.json result
   of the targeted openHAB server (the ID in the URL), however, testing a
   request fires a request to the randomly selected openHAB server
 * Apps (iOS/Android) will always use the randomly selected openHAB instance
   for now
* IFTTT will always choose the randomly selected openHAB instance

However, I'm not sure, how we, on the side of openHAB-cloud, could solve these issues (I think we can't). The best thing we can provide is a working backwards compatible functionality, like we've already: If an openHAB resource is requested without having the OH ID in the request URL, one of the registered OH instances will be selected to serve the request.

Another thing where I don't have any idea how this can be worked around is the response body rewrite based on a regular expression. The background:

OpenHAB currently seems to require to be installed and served on the root of the current domain (or IP address). That means, that some (not all) URL's returned in the body of, e.g., the Basic UI contains links which are relative to the root of the current page. That requires the URL rewrite, unfortunately. If the URLs would always be relative (which would probably require things like `../rest/items...` instead of `/rest/items...`), the rewrite wouldn't be needed, I think. However, maybe someone else has a good idea how this can be solved in openHAB-cloud :]

**Comment by jnherm** [comment]
Hope this will be available soon! Thanks to all contributors!

**Comment by kirantpatil** [comment]
Hi @FlorianSW,

Any updates regarding this feature ?

We will test it and report issues, please let us know how to enable/access it.

Thanks.

**Comment by digitaldan** [comment]
Closing for now

**Comment by tlc76** [comment]
Hi all,

Are there any updates related to this PR? I need to deploy two separate instances of openhab2 in two physical locations, installed on two raspberry pi's, but visible/manageable under the same myopenhab cloud account.

Thanks!

**Comment by FlorianSW** [comment]
Hi,

sorry for not replying to this earlier. The biggest problem why this PR get stalled is because of the things mentioned in https://github.com/openhab/openhab-cloud/pull/98#issuecomment-354882842. The current state of the PR may or may not work already (if someone wants to try it out). However, the implementation is not as good as it needs to be to be used in a production environment.

While I'm pretty ok with the current backward compatibility solution (which randomly selects the first OH instance it can find associated with the user account when requested without the OH instance ID), I'm fairly dissatisfied with the way how the absolute URL problem is handled. This is (as of the current state of the PR) done by rewriting the body of the targeted OH instance. This is really nothing what a proxy should do, as it changes how the content looks like, how it is transported and is simply a fuzzy logic, which may fail in the future. Having something like this in a prod system is always a separate point of failure, which should be avoided.

However, I, as of now, don't have an idea of how this could be solved. It might be possible to add a "X-Base-Href" header (like what the `base` tag in HTML does), however, there's no internet standard for that, so this needs to be well thought and be a defined behaviour before it can be implemented.

Last but not least: When I started implementing this feature, the community did not really longed for it, which means, that putting a lot of effort into the feature did not seemed to be reasonable. I'm not entirely sure if that already changed and if I may have enough time to do the work I mentioned above (which also includes implementing things on the OH side as well). So, the state of the PR might stay as it is for now :(

Best,
Florian

**Pull request closed without merging**